source/analysis/AM1DCalibrationAB.cpp

```diff
@@ -208,113 +219,171 @@ void AM1DCalibrationAB::setInputSources()
-			postEdgePointValue = postEdgePoint();
-
-		// determine value of desired offset reference point
-		double offset = int(data_->value(preEdgePointValue));
-		// shift spectra so that pre edge reference point is 0
-			retVal = retVal - offset;
-
-		// determine value of post edge scaling reference point
-		double scale = data_->value(postEdgePointValue);
-		// scale spectra so that post edge reference point is 1
-		if (scale != 0){ // if the postEdgePoint is zero, do nothing, otherwise scale and shift the output values
-			retVal = retVal / scale;
-
-		}
-	}
-	return retVal;
+        if (isTransmission()){
+
+            if (data.at(preEdgePointValue) == 0) // don't divide by zero
+                offset = 0;
+
+            else if (normalizer.at(preEdgePointValue) * data.at(preEdgePointValue) < 0)  //Don't take the log of a negative number
+                offset = 0;
+
+            else
+                offset = qLn(normalizer.at(preEdgePointValue)/data.at(preEdgePointValue));
+        }
+
+        else{
+
+            if (normalizer.at(preEdgePointValue) == 0)  // don't divide by zero
+                offset = 0;
+
+            else
+                offset = data.at(preEdgePointValue)/normalizer.at(preEdgePointValue);
+        }
+
+        // shift spectra so that pre edge reference point is 0
+        for (int i = 0; i < totalSize; i++)
+            cachedData_[i] -= offset;
+
+        // determine value of post edge scaling reference point, this is ugly because it must handle the case that the reference point is outside the requested index range for all modes...
+        double scale = 1;
+
+        if (isTransmission()){
+
+            if (data.at(postEdgePointValue) == 0) // don't divide by zero
+                scale = 1;
+
+            else if (normalizer.at(postEdgePointValue) * data.at(postEdgePointValue) < 0)  //Don't take the log of a negative number
+                scale = 1;
+
+            else
+                scale = qLn(normalizer.at(postEdgePointValue)/data.at(postEdgePointValue)) - offset;
+        }
+
+        else{
+
+            if (normalizer.at(postEdgePointValue) == 0)  // don't divide by zero
+                scale = 1;
+
+            else
+                scale = (data.at(postEdgePointValue)/normalizer.at(postEdgePointValue)) - offset;
+        }
+
+        // scale spectra so that post edge reference point is 1
+        if (scale != 0){ // if the postEdgePoint is zero, do nothing, otherwise scale and shift the output values
+
+            for (int i = 0; i < totalSize; i++)
+                cachedData_[i] /= scale;
+        }
+    }
+
+    cacheUpdateRequired_ = false;
 }
 
+bool AM1DCalibrationAB::canAnalyze(const QString &dataName, const QString &NormalizationName) const
+{
+	// Can always analyze two 1D data sources.
+	if (sources_.count() == 2)
+		return true;
 
+	// The first data source is reserved for the data.
+	if (indexOfInputSource(dataName) >= 0 && indexOfInputSource(NormalizationName))
+		return true;
 
```

Comment:
Unfortunately, that is a case of unclear commenting.  The logic is correct with the ` >= 0` because the _data_ source can be any data source that has been provided (i.e.: `0` to `size()-1`).